### ISSUE TITLE
Docker 1.9.1 - port forwarding stops working randomly, docker-proxy conn reset

### ISSUE THREAD

**Issue by Soulou** [docker/docker#18575]
I've a container running since a few days:

```
f99fdcc4f55e        custom/mongo:2.6.10                                               "/mongo name-prod"   5 days ago          Up 5 days           0.0.0.0:30053->27017/tcp                                         name9304-1449269114
```

Everything was alright when suddenly all the applications connected to the mongo instance stopped being able to reach the database.

Netcat returned directly the prompt

```
# netcat localhost 30053 -v
Connection to localhost 30053 port [tcp/*] succeeded!
#
```

Let's have a look to tcpdump when running netcat:

```
# tcpdump -i lo tcp port 30053
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on lo, link-type EN10MB (Ethernet), capture size 65535 bytes
18:58:57.296579 IP6 localhost.47307 > localhost.30053: Flags [S], seq 2324257876, win 43690, options [mss 65476,sackOK,TS val 175793459 ecr 0,nop,wscale 7], length 0
18:58:57.296609 IP6 localhost.30053 > localhost.47307: Flags [S.], seq 2490459860, ack 2324257877, win 43690, options [mss 65476,sackOK,TS val 175793459 ecr 175793459,nop,wscale 7], length 0
18:58:57.296622 IP6 localhost.47307 > localhost.30053: Flags [.], ack 1, win 342, options [nop,nop,TS val 175793459 ecr 175793459], length 0
18:58:57.297024 IP6 localhost.30053 > localhost.47307: Flags [F.], seq 1, ack 1, win 342, options [nop,nop,TS val 175793459 ecr 175793459], length 0
18:58:57.297091 IP6 localhost.47307 > localhost.30053: Flags [F.], seq 1, ack 2, win 342, options [nop,nop,TS val 175793459 ecr 175793459], length 0
18:58:57.297100 IP6 localhost.30053 > localhost.47307: Flags [.], ack 2, win 342, options [nop,nop,TS val 175793459 ecr 175793459], length 0
```

From  this log, we can see that instantly the connection is finished by docker-proxy (which has bound the port 30053)

Inside the container, everything is alright:

```
# docker exec -it f99fdcc4f55e bash
mongodb@f99fdcc4f55e:/$ mongo -u user -p password localhost:27017/foobar
MongoDB shell version: 2.6.10
connecting to: localhost:27017/foobar
2015-12-10T18:00:50.950+0000 Error: 18 { ok: 0.0, errmsg: "auth failed", code: 18 } at src/mongo/shell/db.js:1287
exception: login failed
```

As we can see inside the container the process is working fine (login error means connection successful)

ps to see docker proxy
```
# ps aux | grep 30053
root     19280  0.0  0.0 177320 23100 ?        Sl   déc.04   0:35 docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 30053 -container-ip 172.17.0.186 -container-port 27017
```

iptables output:
```
# iptables -t nat -L -n -v | grep 3005
  854 51240 DNAT       tcp  --  !docker0 *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30053 to:172.17.0.186:27017
# iptables -t nat -L -n -v | grep 172.17.0.186
    0     0 MASQUERADE  tcp  --  *      *       172.17.0.186         172.17.0.186         tcp dpt:27017
# iptables -L -n -v | grep 172.17.0.186
90539   11M ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.186         tcp dpt:27017
```

When I try to reach directly the IP: of the container
```
# nc 172.17.0.186 27017 -v
nc: connect to 172.17.0.186 port 27017 (tcp) failed: Connection refused
```

Now I compare to a working container on the same host, it works all right.
```
# nc 172.17.0.59 20001 -v
Connection to 172.17.0.59 20001 port [tcp/*] succeeded!
^C
```

So I'm looking for hints, ideas, solutions to this problem, what could have led to that?
Docker restart doesn't help, I've to stop, remove and recreate the container to fix that.

```
# docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:12:04 UTC 2015
 OS/Arch:      linux/amd64
```

```
# docker info
Containers: 229
Images: 542
Server Version: 1.9.1
Storage Driver: btrfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.0-19-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 6
Total Memory: 31.42 GiB
Name: [FILTERED]
ID: AZNH:RYE6:EPU2:OM3H:RQGL:E6GL:3PCA:BJJP:JVQO:ILYS:G7B4:TWO
Registry: https://index.docker.io/v1/
```

**Comment by Soulou** [comment]
Actually stopping the container, removing it and recreating it again has no effect :(

**Comment by Soulou** [comment]
After restarting the container I get multiple entries in iptables, maybe a hint:

FIrst is the old one, second is the newly created container

```
171   2193  219K ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.172         tcp dpt:20070
172      0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.177         tcp dpt:20071
173   1985  161K ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.175         tcp dpt:20067
174      0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.176         tcp dpt:20074
175    371  133K ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.191         tcp dpt:20072
176      0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.180         tcp dpt:27017
177      0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.187         tcp dpt:20040
178     20  1200 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.172         tcp dpt:27017

    0     0 MASQUERADE  tcp  --  *      *       172.17.0.172         172.17.0.172         tcp dpt:20070
    0     0 MASQUERADE  tcp  --  *      *       172.17.0.172         172.17.0.172         tcp dpt:27017

  159  9540 DNAT       tcp  --  !docker0 *       0.0.0.0/0            0.0.0.0/0            tcp dpt:20070 to:172.17.0.172:20070
   33  1980 DNAT       tcp  --  !docker0 *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30053 to:172.17.0.172:27017
```

**Comment by Soulou** [comment]
According to `docker network inspect` several containers got the same IP:

```
~# docker network inspect bridge | grep -A 5 -B 5 '0\.172'
            "732569d2a0409a34b12fcd136b419b621e1e5cbea38faa0e4f2a853f9f17be13": {
                "EndpointID": "b4d0d4a131441842cfcf6c66342fd4d235adedc1745b40a1454e87d9fa4bacef",
                "MacAddress": "02:42:ac:11:00:ac",
                "IPv4Address": "172.17.0.172/16",
                "IPv6Address": ""
            },
--
            "c2afa171bbadcb637088ef10b05851503f500d2a61dad80d98dc5fe9b943f423": {
                "EndpointID": "adfd47a5ee5169676205f91bc20a55b9ffecd3c709732c186dfd85775937493c",
                "MacAddress": "02:42:ac:11:00:ac",
                "IPv4Address": "172.17.0.172/16",
                "IPv6Address": ""
            }
```

**Comment by Soulou** [comment]
Ok it seems that is the problem, docker is re-allocating IPs, the new iptables rules seem to mask the those from the first container with the IP, making it unreachable.

**Comment by Soulou** [comment]
Maybe related to https://github.com/docker/docker/issues/18535

**Comment by Soulou** [comment]
Closed in favor of #18535
